### ISSUE TITLE
Type in non-public module leaks out to public interface

### ISSUE THREAD

**Issue by faern** [rust-lang/rust#38253]
If I have a public type in a non-public module that type is not part of the crate's public interface. However, if I use this type in the public interface of the crate rustc does not complain and the crate compiles.

In this state it will not be possible to call that function, since the type can't be used. Also the documentation looks strange as the private type will be listed as part of the public interface, but it won't be a link to the page about that type, since there is no such page.

Code producing the problem:

```rust
mod secrets {
    pub struct Secret;
}

pub fn fail(_: secrets::Secret) {}
```

I expect this to happen: Compiler error similar to `can't leak private type`

## Meta
The problem exists on both stable and nightly. These versions:
```
rustc 1.13.0 (2c6933acc 2016-11-07)
binary: rustc
commit-hash: 2c6933acc05c61e041be764cb1331f6281993f3f
commit-date: 2016-11-07
host: x86_64-unknown-linux-gnu
release: 1.13.0
```
```
rustc 1.15.0-nightly (daf8c1dfc 2016-12-05)
binary: rustc
commit-hash: daf8c1dfce3b448fc581cc319f64632ec22bd0e1
commit-date: 2016-12-05
host: x86_64-unknown-linux-gnu
release: 1.15.0-nightly
LLVM version: 3.9
```

**Comment by petrochenkov** [comment]
This behavior has a long history, but this is not a bug and the code works as expected.
[RFC 136](https://github.com/rust-lang/rfcs/blob/master/text/0136-no-privates-in-public.md) describes the rules (more or less).
The rule of thumb is that if some entity is marked with `pub` then it's considered public for all purposes even if you can't always actually name it.

**Comment by faern** [comment]
Got it. Strange behavior, but now I know.
